platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/sync/protocol/SyncProtocol.java

```diff
@@ -0,0 +1,234 @@
+     */
+    private boolean peerNeededForFallenBehind() {
+        final List<Long> neededForFallenBehind = fallenBehindManager.getNeededForFallenBehind();
+
+        return neededForFallenBehind != null && neededForFallenBehind.contains(peerId.getId());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean shouldInitiate() {
+        // are there any reasons not to initiate?
+        if (permit.isLockAcquired() || !peerAgnosticSyncChecks.shouldSync() || fallenBehindManager.hasFallenBehind()) {
+            return false;
+        }
+
+        // is there a reason to initiate?
+        if (peerNeededForFallenBehind() || criticalQuorum.isInCriticalQuorum(peerId.getId())) {
+            permit = outgoingPermitProvider.tryAcquire();
+
+            return permit.isLockAcquired();
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean shouldAccept() {
+        // are there any reasons not to accept?
+        if (permit.isLockAcquired() || !peerAgnosticSyncChecks.shouldSync() || fallenBehindManager.hasFallenBehind()) {
+            syncMetrics.updateRejectedSyncRatio(true);
+            return false;
+        }
+
+        permit = incomingPermitProvider.tryAcquire();
+        final boolean isLockAcquired = permit.isLockAcquired();
+
+        syncMetrics.updateRejectedSyncRatio(!isLockAcquired);
+
+        return isLockAcquired;
+    }
+
+    /**
+     * Closes the existing permit, and resets the {@link #permit} member variable
+     */
+    private void closePermit() {
+        permit.close();
+        permit = MaybeLocked.NOT_ACQUIRED;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void initiateFailed() {
+        closePermit();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean acceptOnSimultaneousInitiate() {
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void runProtocol(@NonNull final Connection connection)
+            throws NetworkProtocolException, IOException, InterruptedException {
+
+        throwArgNull(connection, "connection");
+
+        if (!permit.isLockAcquired()) {
+            throw new NetworkProtocolException("sync permit not acquired prior to executing sync protocol");
+        }
+
+        try {
+            synchronizer.synchronize(connection);
+        } catch (final ParallelExecutionException | SyncException e) {
+            if (Utilities.isRootCauseSuppliedType(e, IOException.class)) {
+                throw new IOException(e);
+            }
+
+            throw new NetworkProtocolException(e);
+        } finally {
+            closePermit();
+
+            final long sleepAfterSyncDuration = sleepAfterSyncSupplier.getAsLong();
+            if (sleepAfterSyncDuration > 0) {
+                Thread.sleep(sleepAfterSyncDuration);
+            }
```

Comment:
Would it be possible to remove this sleep, and instead have the protocol `shouldAccept()` return false if a sync has happened too recently?

A useful helper class is `com.swirlds.common.utility.throttleRateLimiter.java`. You can tell it the maximum number of times something should happen per second, and it will tell you if not enough time has passed.


Example:

```
final RateLimiter rateLimiter = new RateLimiter(time, 10 /* permit syncing at 10 hz */);

// ...

    public boolean shouldAccept() {
        // are there any reasons not to accept?
        if (permit.isLockAcquired() || !peerAgnosticSyncChecks.shouldSync() || fallenBehindManager.hasFallenBehind()) {
            syncMetrics.updateRejectedSyncRatio(true);
            return false;
        }

        permit = incomingPermitProvider.tryAcquire();
        final boolean isLockAcquired = permit.isLockAcquired();

        syncMetrics.updateRejectedSyncRatio(!isLockAcquired);

        if (isLockAcquired && !rateLimiter.request()) {
                closePermit();
                return false;
        }

        return isLockAcquired;
    }
```

Note that this will not be exactly logically consistent with the previous implementation. The old setting for the amount of time to sleep after a sync will instead become the minimum time between the start of sequential syncs. I don't expect that to cause problems as long as we carefully document it.

Reply:
@cody-littley I think this would be fine, but it's made more complicated by the fact that the time-to-sleep can change.

So we couldn't just set up the rate limiter, we'd have to get the new time-to-sleep after each sync, compare against the old time-to-sleep, and maybe make a new limiter if it's changed.

Alternatively, I could home-roll a timer to accomplish the rate limiting,

Reply:
After further thought, let's not make this change now. Simpler to use the old pattern (even if it is silly) than to test code that we hope to deprecate after chatter is enabled.